packages/vite/src/node/plugins/html.ts

```diff
@@ -344,22 +344,40 @@ export function buildHtmlPlugin(config: ResolvedConfig): Plugin {
 
         // inject chunk asset links
         if (chunk) {
-          const assetTags = [
+          const assetTags = []
+
+          if (
+            chunk.imports.length ||
+            Object.entries(chunk.modules).some(
+              ([id, mod]) =>
+                id !== 'vite/dynamic-import-polyfill' &&
+                mod.renderedLength &&
+                !this.getModuleInfo(id)?.meta.css
+            )
+          ) {
             // js entry chunk for this page
-            {
+            assetTags.push({
               tag: 'script',
               attrs: {
                 type: 'module',
                 crossorigin: true,
                 src: toPublicPath(chunk.fileName, config)
               }
-            },
+            })
+          } else {
+            // entry js chunk was effectively empty, prevent rollup outputting it
+            delete bundle[chunk.fileName]
```

Comment:
It feels weird to be mutating the bundle in the HTML plugin - maybe this should be done somewhere else (a separate plugin?) so that the HTML plugin's logic remains intact.

Reply:
I could be wrong but I think that would make the implementation a bit more awkward, but probably doable. I think I'd need to extract https://github.com/vitejs/vite/pull/3129/files#diff-89bae1df62862bb7f4a03d82a1e9cbf4ac6d0c042f21fbbacb0a2238bd050042R350-R356 because it would have to be called in two places. Once here in the plugin, since the plugin still needs to know if it should output the script, and once in this other plugin.

The problem is we can't exactly delete the chunk from the bundle until after the html plugin has looked at it, since it still needs to analyze preload/stylesheet links for that chunk even if there is no js.